web/v1/snowflake/sql-runner/sql/standard/00-setup/01-main/01-stored-procedures.sql

```diff
@@ -156,139 +160,111 @@ CREATE OR REPLACE PROCEDURE {{.output_schema}}.column_check(SRC_SCHEMA  VARCHAR,
+  DERIVED_PARTITION_KEY: the partition key for derived table
+  UPSERT_LOOKBACK FLOAT: corresponds to {{.upsert_lookback_days}}::float
   To drop:
-  DROP PROCEDURE {{.output_schema}}.commit_with_metadata(VARCHAR,VARCHAR,VARCHAR,VARCHAR,
-                                                         VARCHAR,VARCHAR,VARCHAR,VARCHAR,
-                                                         VARCHAR,VARCHAR,VARCHAR,VARCHAR,
-                                                         VARCHAR,VARCHAR,VARCHAR,FLOAT);
-
+  DROP PROCEDURE {{.output_schema}}.commit_tables(VARCHAR,VARCHAR,VARCHAR,VARCHAR,
+                                                  VARCHAR,VARCHAR,VARCHAR,VARCHAR,
+                                                  VARCHAR,VARCHAR,FLOAT);
 */
-CREATE OR REPLACE PROCEDURE {{.output_schema}}.commit_with_metadata(
-                                                        META_TBL            VARCHAR,
-                                                        META_FROM_TBL       VARCHAR,
-                                                        STAGE_NXT           VARCHAR,
-                                                        STG_SRC_SCHEMA      VARCHAR,
-                                                        STG_SRC_TBL         VARCHAR,
-                                                        STG_TRG_SCHEMA      VARCHAR,
-                                                        STG_TRG_TBL         VARCHAR,
-                                                        STG_JOIN_KEY        VARCHAR,
-                                                        SKIP_DRV            VARCHAR,
-                                                        DRV_SRC_SCHEMA      VARCHAR,
-                                                        DRV_SRC_TBL         VARCHAR,
-                                                        DRV_TRG_SCHEMA      VARCHAR,
-                                                        DRV_TRG_TBL         VARCHAR,
-                                                        DRV_JOIN_KEY        VARCHAR,
-                                                        DRV_PARTITION_KEY   VARCHAR,
-                                                        DRV_UPSERT_LOOKBACK FLOAT)
+CREATE OR REPLACE PROCEDURE {{.output_schema}}.commit_tables(
+                                                 STAGE_NEXT            VARCHAR,
+                                                 AUTOMIGRATE           VARCHAR,
+                                                 STAGING_SOURCE        VARCHAR,
+                                                 STAGING_TARGET        VARCHAR,
+                                                 STAGING_JOIN_KEY      VARCHAR,
+                                                 SKIP_DERIVED          VARCHAR,
+                                                 DERIVED_SOURCE        VARCHAR,
+                                                 DERIVED_TARGET        VARCHAR,
+                                                 DERIVED_JOIN_KEY      VARCHAR,
```

Comment:
Join key and partition key should always be the same across all 3 tables involved. So each only need to be one input.

Reply:
@colmsnowplow just to confirm, could you please explain what do you expect to be the same? Do you maybe mean stage-joinKey Vs production-joinKey? Which 3 tables do you refer to?

Reply:
So for example in the page views module we have three tables:

`page_views_this_run`, `page_views_staged`, `page_views`. These by definition have the exact same definition, including joinkey and partition key. 

So the function which handles their commit needs only one respective `paritionKey` and `joinKey` argument. If these are ever not the same, something is wrong and we shouldn't allow it. Make sense?

Reply:
Thank you very much for explaining. This is indeed the case for the page-views module, but in sessions module there are 4 tables, where there is a difference:

- the join key for `sessions_this_run` - `sessions` is `domain_sessionid`
- the join key for `sessions_userid_manifest_this_run` - `sessions_userid_manifest_staged` is `domain_userid`

So for a procedure to abstract over all, different arguments seem needed.

Reply:
Makes sense. Seems to be another reason to prefer this function to handle one update at a time.

Reply:
Just to clarify here, which also relates to the comment on performance tests with splitting the transactions:
For both redshift and bigquery, those statements are inside a transaction. Maybe until this is changed there too (if not really needed) it makes sense to have the same logic for snowflake too, at least for v1.0.0. This was the reason for combining those statements in the same stored procedure (and this has also been the reason for many of the superficial differences so far).
Also, since Snowflake allows to overload a procedure name, usability is not really an issue in the long run.
I would prefer so instead of deviating at this point.

Reply:
I'm not sure I've made the considerations v clear as I was still thinking about it when we spoke. Let me try to do so:

The best design for our functions (in this case a procedure) is the simplest. One unit of work at a time, and as generalisable as possible. 

We can see the compromise we've made here impact other parts of the model's design already (in other comments).

I think we also expect a lot of users to follow how to use these - most of them won't understand that the underlying models use two separate pre-production tables to stage updates, so it is not intuitive to them what all of the arguments are for.

These are more than just superficial considerations for me.

I see the decision here as a choice between sacrificing the best design, or sacrificing consistency in how transactions are done.

Nesting the transactions has very little (if any) actual impact on the model, as long as DELETE and INSERT are in the same transaction. We just do it in the other DBs because we can.

So for me, consistency in this aspect doesn't really outweigh the benefits of a simpler design.

Now, if performing two transactions does make a significant difference to performance, or if there is another material difference I'm missing, then we should definitely consider those things.

Does this make sense?

Reply:
> We can see the compromise we've made here impact other parts of the model's design already (in other comments).

It seems this is still under discussion.

> I see the decision here as a choice between sacrificing the best design, or sacrificing consistency in how transactions are done.

In my mind there is no need to sacrifice anything here. Snowflake allows for more options on this, so instead of one or the other, i see a potential to achieve both. Would you consider such options?

Reply:
Of course if we can avoid both problems I would prefer that. What solution do you propose?

Reply:
Thank you for considering this. So, the option is to define another stored procedure that does exactly as you propose, i.e. do one thing at the time. It can actually have the same name with just diff number of arguments, as Snowflake allows for this overload.
Advantages:
1. Not yet another procedure. Same name that allows for both use cases.
2. Keeping the same to other models' logic on the commit steps.
3. Easiest to change. The only thing is just to add this definition. Nothing else needs to change on the model, so the quickest at this point for v1.0.0
Disadvantages:
1. Some code duplication, as we overload.

I guess there may be more disadvantages as i don't have the experience to judge the quality of this design, but is indeed a quick change that at this point allows to have the simpler design you mention(and i agree) for any simpler use case and not having to sacrifice.
If you still think it is best design to do otherwise, i will go on to do so, but i estimate some more time will be needed.

Reply:
I don't like duplicating the code, and I'm not sure if we're actually saving time by doing it this way - once you have a new function surely it's very little effort to use it in the SQL. 

However if you feel that's the best way forward then sure - we can clean up what we need to in a follow-up release.

Reply:
Thank you @colmsnowplow. I understand this overload option is not ideal, so i made the changes you suggest in the commit steps (the performance diff was around 3sec for 1,700,000 events, so i assume insignificant as argument to keep in same transaction). These changes were also force-pushed in "Address review changes III".